sql/api/src/main/scala/org/apache/spark/sql/types/TimeType.scala

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.types
+
+import org.apache.spark.annotation.Unstable
+import org.apache.spark.sql.errors.DataTypeErrors
+
+/**
+ * The time type represents a time value with fields hour, minute, second, up to microseconds. The
+ * range of times supported is 00:00:00.000000 to 23:59:59.999999.
+ *
+ * @param precision
+ *   The time fractional seconds precision which indicates the number of decimal digits maintained
+ *   following the decimal point in the seconds value. The supported range is [0, 6].
+ *
+ * @since 4.1.0
+ */
+@Unstable
+case class TimeType(precision: Int) extends DatetimeType {
+
+  if (precision < 0 || precision > 6) {
+    throw DataTypeErrors.unsupportedTimePrecisionError(precision)
+  }
+
+  /**
+   * The default size of a value of the TimeType is 8 bytes.
+   */
+  override def defaultSize: Int = 8
```

Comment:
an integer should be 4 bytes?

Reply:
@cloud-fan For `Integer` yes, but I had to use `Long` to comprise all values with microsecond precision: 24 * 60 * 60 * 1000 * 1000 = 86400000000 > Int.MaxValue. We can consider allocation of `Integer` for smaller precision, but for now I would propose to always allocate `Long`.